# Prova 2 - Parte 2

## SME0823 Modelos de Regressão e Aprendizado Supervisionado II

#### Andrey Inácio Paulino 10818741

### Leitura e descrição dos dados

Importação da biblioteca pandas para leitura dos dados

In [ ]:
import pandas as pd
dados = pd.read_csv("C:\\Users\\inacio\\Desktop\\oitavo_semestre\\modelos_regressao_II\\coverageX.txt", delimiter = "\t")

In [ ]:
dados

y  MEN  URBAN  PRIVATE  AGE  SENIORITY
0     0    1      1        1   58         32
1     1    1      0        1   37         30
2     1    1      0        1   56         28
3     0    1      0        1   68         27
4     0    1      0        1   58         26
...  ..  ...    ...      ...  ...        ...
3995  0    0      0        1   30          2
3996  1    0      0        1   23          2
3997  1    0      1        1   41          2
3998  1    0      1        1   27          2
3999  0    1      1        1   41          2

[4000 rows x 6 columns]

In [ ]:
dados['y'].value_counts()

0    2613
1    1387
Name: y, dtype: int64

No conjunto de dados coverageX.txt a variável resposta é sobre acionar ou não o seguro de carros, a base de dados possui 4000 observações e seis colunas.
Os dados foram utilizados em Bazán et al (2017) e são descritos a seguir:
* Y: Acionou (1) ou não (0) o seguro do carro;
* MEN: Home (1), Mulher (0);
* URBAN: Conduz na área urbana (1), conduz na área rural (0);
* PRIVATE: Veículo privado (1), veículo público (0);
* AGE: Idade em anos do condutor;
* SENIORITY: Tempo de uso da carteira de motorista.

Sendo que das 4000 observações, 1387 o seguro foi acionado.

In [ ]:
dados.dropna()

y  MEN  URBAN  PRIVATE  AGE  SENIORITY
0     0    1      1        1   58         32
1     1    1      0        1   37         30
2     1    1      0        1   56         28
3     0    1      0        1   68         27
4     0    1      0        1   58         26
...  ..  ...    ...      ...  ...        ...
3995  0    0      0        1   30          2
3996  1    0      0        1   23          2
3997  1    0      1        1   41          2
3998  1    0      1        1   27          2
3999  0    1      1        1   41          2

[4000 rows x 6 columns]

A base de dados não apresenta valores faltantes.

#### a) Use seu código USP para tomar uma amostra de tamanho 2000 para desenvolver as análises que são descritas a seguir. Salve esta base com o nome baseprincipal.csv. Este base de dados deverá ser compartilhado com seu relatório.

Pegando amostra de tamanho 2000 do conjunto de dados e salvando a amostra como uma nova base de dados no formato csv.

In [ ]:
amostra = dados.sample(2000, random_state = 10818741)
amostra.to_csv('baseprincipal.csv',sep=',',decimal='.')
amostra

y  MEN  URBAN  PRIVATE  AGE  SENIORITY
1487  0    1      0        1   41         12
3440  0    1      0        1   49         15
1206  0    1      1        1   66         23
2071  0    1      0        1   37         28
3572  0    0      1        1   18          3
...  ..  ...    ...      ...  ...        ...
1508  1    1      0        1   44         20
2330  0    1      0        1   43          7
203   0    1      0        1   36         17
1722  0    1      0        1   45         11
3618  0    1      0        1   64          3

[2000 rows x 6 columns]

#### b) Crie uma amostra de treino e teste (70% treino e 30% teste) usando a base de dados principal criada no passo a). Salve estas bases como o nome basetreino.csv e baseteste.csv. Estas bases de dados deveram ser compartilhadas com seu relatório.

Importação da biblioteca numy para trabalhar com arrays

In [ ]:
import numpy as np

Divisão dos dados

In [ ]:
X = amostra[amostra.columns[1:]]
X

MEN  URBAN  PRIVATE  AGE  SENIORITY
1487    1      0        1   41         12
3440    1      0        1   49         15
1206    1      1        1   66         23
2071    1      0        1   37         28
3572    0      1        1   18          3
...   ...    ...      ...  ...        ...
1508    1      0        1   44         20
2330    1      0        1   43          7
203     1      0        1   36         17
1722    1      0        1   45         11
3618    1      0        1   64          3

[2000 rows x 5 columns]

In [ ]:
y = amostra['y']
y

1487    0
3440    0
1206    0
2071    0
3572    0
       ..
1508    1
2330    0
203     0
1722    0
3618    0
Name: y, Length: 2000, dtype: int64

Conversão de X e y em array para criar as amostras de treino e teste

In [ ]:
treino = X.values 
treino

array([[ 1,  0,  1, 41, 12],
       [ 1,  0,  1, 49, 15],
       [ 1,  1,  1, 66, 23],
       ...,
       [ 1,  0,  1, 36, 17],
       [ 1,  0,  1, 45, 11],
       [ 1,  0,  1, 64,  3]], dtype=int64)

In [ ]:
teste = np.array(y)
teste

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

Importação da biblioteca e criação das amostras de treino e teste

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_treino,X_teste,y_treino,y_teste = train_test_split(treino, teste, test_size=0.3, random_state=10818741) 

Ajuste dos formatos array e junção das partes treino e teste para depois converter data frame

In [ ]:
y_treino = np.array_split(y_treino, len(y_treino))
base_treino = np.concatenate((X_treino, y_treino), axis=1)

In [ ]:
base_treino

array([[ 1,  0,  1, 37,  3,  0],
       [ 1,  1,  1, 63, 13,  0],
       [ 1,  0,  1, 48, 20,  1],
       ...,
       [ 0,  1,  1, 44, 18,  0],
       [ 1,  1,  1, 51,  7,  0],
       [ 1,  0,  1, 36, 16,  1]], dtype=int64)

In [ ]:
y_teste = np.array_split(y_teste, len(y_teste)) 
base_teste = np.concatenate((X_teste, y_teste), axis=1) 

In [ ]:
base_teste

array([[ 1,  0,  1, 44,  2,  0],
       [ 1,  0,  1, 57, 23,  1],
       [ 1,  0,  1, 65, 24,  0],
       ...,
       [ 1,  0,  1, 54, 11,  0],
       [ 0,  0,  1, 31,  4,  1],
       [ 1,  0,  1, 46, 20,  1]], dtype=int64)

Conversão em data frame para salvar a base em formato csv.

In [ ]:
base_treino = pd.DataFrame(base_treino, columns=['MEN', 'URBAN', 'PRIVATE', 'AGE', 'SENIORITY','y'])
base_teste = pd.DataFrame(base_teste, columns=['MEN', 'URBAN', 'PRIVATE', 'AGE', 'SENIORITY','y'])

In [ ]:
base_treino

MEN  URBAN  PRIVATE  AGE  SENIORITY  y
0       1      0        1   37          3  0
1       1      1        1   63         13  0
2       1      0        1   48         20  1
3       0      1        1   53          5  0
4       1      0        1   33         16  1
...   ...    ...      ...  ...        ... ..
1395    1      0        1   45          6  0
1396    1      0        1   41         14  0
1397    0      1        1   44         18  0
1398    1      1        1   51          7  0
1399    1      0        1   36         16  1

[1400 rows x 6 columns]

In [ ]:
base_teste

MEN  URBAN  PRIVATE  AGE  SENIORITY  y
0      1      0        1   44          2  0
1      1      0        1   57         23  1
2      1      0        1   65         24  0
3      0      1        1   34          9  0
4      1      0        1   51         16  0
..   ...    ...      ...  ...        ... ..
595    1      0        1   72         11  0
596    1      0        1   65         14  0
597    1      0        1   54         11  0
598    0      0        1   31          4  1
599    1      0        1   46         20  1

[600 rows x 6 columns]

In [ ]:
base_treino.to_csv('basetreino.csv',sep=',',decimal='.') 
base_teste.to_csv('baseteste.csv',sep=',',decimal='.')

#### c) Desenvolva uma análise completa com os dados de treino seguindo a metodologia da análise de dados de Spam apresentada na aula aberta (ver arquivo SpamAnalysisSlidesHorizontal2021.pdf) (considerando uma determinada função de ligação, escolhendo as variáveis e avaliando a retirada de pontos). 

Importação das bibliotecas para realizar ajuste de modelos de regressão

In [ ]:
import statsmodels.api as sm 
import statsmodels.formula.api as smf 

### Primeiro será ajustado o modelo de regressão binária usando a função de ligação logito(canônica)

In [ ]:
ajuste1 = smf.glm(formula='y ~ MEN + URBAN + PRIVATE + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.logit),
                  data=base_treino).fit() 
print(ajuste1.summary()) 

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1394
Model Family:                Binomial   Df Model:                            5
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -770.82
Date:                Thu, 09 Dec 2021   Deviance:                       1541.6
Time:                        23:23:21   Pearson chi2:                 1.41e+03
No. Iterations:                     5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.2777      0.736      0.377      0.7

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


Para um nível de significância de $\alpha=5\%$, temos a hipóteste $H_0: \beta_j=0, j=0,...,5$ vs $H_1: Para \space qualquer \space B_j \neq 0$.

Então considerando o nível de 0.05, temos que o intercepto e a variável private não são significativos para o modelo. Isso também é observado através dos intervalos de confiança pois tanto o intercepto como a variável private contém o valor zero em seus intervalos de confiança.




#### Modelo sem a variável private

In [ ]:
ajuste12 = smf.glm(formula='y ~ MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.logit),
                  data=base_treino).fit() 
print(ajuste12.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -770.98
Date:                Thu, 09 Dec 2021   Deviance:                       1542.0
Time:                        23:23:21   Pearson chi2:                 1.41e+03
No. Iterations:                     5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.6572      0.242      2.718      0.0

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


Para o modelo sem a variável private, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Modelo sem intercepto

In [ ]:
ajuste13 = smf.glm(formula='y ~ -1 + MEN + URBAN + PRIVATE + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.logit),
                  data=base_treino).fit() 
print(ajuste13.summary())

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -770.89
Date:                Thu, 09 Dec 2021   Deviance:                       1541.8
Time:                        23:23:21   Pearson chi2:                 1.41e+03
No. Iterations:                     5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.8436      0.142     -5.949      0.0

Para o modelo sem intercepto, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Modelo sem intercepto e variável private

In [ ]:
ajuste14 = smf.glm(formula='y ~ -1 + MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.logit),
                  data=base_treino).fit() 
print(ajuste14.summary())

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1396
Model Family:                Binomial   Df Model:                            3
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -774.71
Date:                Thu, 09 Dec 2021   Deviance:                       1549.4
Time:                        23:23:21   Pearson chi2:                 1.39e+03
No. Iterations:                     4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.7460      0.138     -5.413      0.0

Para o modelo sem intercepto e sem a variável private, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Comparando o AIC dos modelos ajustados

In [ ]:
print('Modelo completo:', ajuste1.aic,'\n',
      'Modelo sem a variável private:',ajuste12.aic,'\n',
      'Modelo sem intercepto:',ajuste13.aic,'\n',
      'Modelo sem intecepto e sem a variável private:',ajuste14.aic)

Modelo completo: 1553.646515589297 
 Modelo sem a variável private: 1551.959270637772 
 Modelo sem intercepto: 1551.7842635655402 
 Modelo sem intecepto e sem a variável private: 1557.4171830551454


Segundo o critério AIC, o melhores modelos são o sem intercepto e o modelo sem a variável private. O modelo tem um AIC menor mas a diferença é muito pequena.

### Agora será ajustado o modelo de regressão binária usando a função de ligação probito

In [ ]:
ajuste2 = smf.glm(formula='y ~ MEN + URBAN + PRIVATE + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.probit),
                  data=base_treino).fit() 
print(ajuste2.summary()) 

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1394
Model Family:                Binomial   Df Model:                            5
Link Function:                 probit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -770.80
Date:                Thu, 09 Dec 2021   Deviance:                       1541.6
Time:                        23:23:21   Pearson chi2:                 1.45e+03
No. Iterations:                     5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.1780      0.431      0.413      0.6

No modelo com ligação probito ocorre o mesmo que o modelo de ligação logito, considerando $\alpha=0.05$, o intercepto e a variável private não são significativos para o modelo.

#### Modelo sem a variável private

In [ ]:
ajuste22 = smf.glm(formula='y ~ MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.probit),
                  data=base_treino).fit() 
print(ajuste22.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                 probit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -770.91
Date:                Thu, 09 Dec 2021   Deviance:                       1541.8
Time:                        23:23:21   Pearson chi2:                 1.45e+03
No. Iterations:                     5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.3685      0.144      2.563      0.0

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


Para o modelo sem a variável private, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Modelo sem intercepto

In [ ]:
ajuste23 = smf.glm(formula='y ~ -1 + MEN + URBAN + PRIVATE + AGE + SENIORITY',
                   family=sm.families.Binomial(sm.families.links.probit),
                  data=base_treino).fit() 
print(ajuste23.summary())

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:2: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  


                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                 probit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -770.88
Date:                Thu, 09 Dec 2021   Deviance:                       1541.8
Time:                        23:23:21   Pearson chi2:                 1.45e+03
No. Iterations:                     5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.4976      0.085     -5.865      0.0

Para o modelo sem intercepto, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Modelo sem intercepto e variável private

In [ ]:
ajuste24 = smf.glm(formula='y ~ -1 + MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.probit),
                  data=base_treino).fit() 
print(ajuste24.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1396
Model Family:                Binomial   Df Model:                            3
Link Function:                 probit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -774.21
Date:                Thu, 09 Dec 2021   Deviance:                       1548.4
Time:                        23:23:21   Pearson chi2:                 1.42e+03
No. Iterations:                     5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.4432      0.082     -5.390      0.0

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


Para o modelo sem intercepto e sem a variável private, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Comparando o AIC dos modelos ajustados

In [ ]:
print('Modelo completo:', ajuste2.aic,'\n',
      'Modelo sem a variável private:',ajuste22.aic,'\n',
      'Modelo sem intercepto:',ajuste23.aic,'\n',
      'Modelo sem intecepto e sem a variável private:',ajuste24.aic)

Modelo completo: 1553.5905332637255 
 Modelo sem a variável private: 1551.8278682750276 
 Modelo sem intercepto: 1551.7671521337288 
 Modelo sem intecepto e sem a variável private: 1556.4191785506787


Como ocorrido no modelo com ligação logito, para os modelos com ligação probito segundo o critério AIC, o melhores modelos são o sem intercepto e o modelo sem a variável private. O modelo sem intercepto tem um AIC menor do que o modelo sem a variável private mas a diferença é muito pequena.

### Agora será ajustado o modelo de regressão binária usando a função de ligação cauchito

In [ ]:
ajuste3 = smf.glm(formula='y ~ MEN + URBAN + PRIVATE + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.cauchy),
                  data=base_treino).fit() 
print(ajuste3.summary()) 

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1394
Model Family:                Binomial   Df Model:                            5
Link Function:                 cauchy   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -777.65
Date:                Thu, 09 Dec 2021   Deviance:                       1555.3
Time:                        23:23:22   Pearson chi2:                 1.34e+03
No. Iterations:                    10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.1358      1.003     -0.135      0.8

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


No modelo com ligação cauchito, considerando $\alpha=0.05$, o intercepto e a variável private não são significativos para o modelo.

#### Modelo sem a variável private

In [ ]:
ajuste32 = smf.glm(formula='y ~ MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.cauchy),
                  data=base_treino).fit() 
print(ajuste32.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                 cauchy   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -778.21
Date:                Thu, 09 Dec 2021   Deviance:                       1556.4
Time:                        23:23:22   Pearson chi2:                 1.33e+03
No. Iterations:                    10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.7941      0.244      3.249      0.0

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


Para o modelo sem a variável private, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Modelo sem intercepto

In [ ]:
ajuste33 = smf.glm(formula='y ~ -1 + MEN + URBAN + PRIVATE + AGE + SENIORITY',
                   family=sm.families.Binomial(sm.families.links.cauchy),
                  data=base_treino).fit() 
print(ajuste33.summary())

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:2: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  


                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                 cauchy   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -777.66
Date:                Thu, 09 Dec 2021   Deviance:                       1555.3
Time:                        23:23:22   Pearson chi2:                 1.33e+03
No. Iterations:                    10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.8834      0.145     -6.072      0.0

Para o modelo sem intercepto, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Modelo sem intercepto e variável private

In [ ]:
ajuste34 = smf.glm(formula='y ~ -1 + MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.cauchy),
                  data=base_treino).fit() 
print(ajuste34.summary())

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1396
Model Family:                Binomial   Df Model:                            3
Link Function:                 cauchy   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -783.20
Date:                Thu, 09 Dec 2021   Deviance:                       1566.4
Time:                        23:23:22   Pearson chi2:                 1.33e+03
No. Iterations:                     9                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.7492      0.139     -5.372      0.0

Para o modelo sem intercepto e sem a variável private, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Comparando o AIC dos modelos ajustados

In [ ]:
print('Modelo completo:', ajuste3.aic,'\n',
      'Modelo sem a variável private:',ajuste32.aic,'\n',
      'Modelo sem intercepto:',ajuste33.aic,'\n',
      'Modelo sem intecepto e sem a variável private:',ajuste34.aic)

Modelo completo: 1567.3039334297305 
 Modelo sem a variável private: 1566.4266600700805 
 Modelo sem intercepto: 1565.322150178149 
 Modelo sem intecepto e sem a variável private: 1574.4058915908183


No modelo com ligação cauchito segundo o critério AIC, o melhor modelo é o sem intercepto.

### Agora será ajustado o modelo de regressão binária usando a função de ligação cloglog

In [ ]:
ajuste4 = smf.glm(formula='y ~ MEN + URBAN + PRIVATE + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.cloglog),
                  data=base_treino).fit() 
print(ajuste4.summary()) 

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1394
Model Family:                Binomial   Df Model:                            5
Link Function:                cloglog   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -783.20
Date:                Thu, 09 Dec 2021   Deviance:                       1566.4
Time:                        23:23:22   Pearson chi2:                 1.37e+03
No. Iterations:                    12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.2965      0.620     -0.478      0.6

No modelo com ligação cloglog, considerando $\alpha=0.05$, o intercepto e a variável private não são significativos para o modelo.

#### Modelo sem a variável private

In [ ]:
ajuste42 = smf.glm(formula='y ~ MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.cloglog),
                  data=base_treino).fit() 
print(ajuste42.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                cloglog   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -783.43
Date:                Thu, 09 Dec 2021   Deviance:                       1566.9
Time:                        23:23:22   Pearson chi2:                 1.37e+03
No. Iterations:                    12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0651      0.184      0.354      0.7

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


Para o modelo sem a variável private, considerando $\alpha=0.05$, o intercepto não é significativo para o modelo.

#### Modelo sem intercepto

In [ ]:
ajuste43 = smf.glm(formula='y ~ -1 + MEN + URBAN + PRIVATE + AGE + SENIORITY',
                   family=sm.families.Binomial(sm.families.links.cloglog),
                  data=base_treino).fit() 
print(ajuste43.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1395
Model Family:                Binomial   Df Model:                            4
Link Function:                cloglog   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -783.33
Date:                Thu, 09 Dec 2021   Deviance:                       1566.7
Time:                        23:23:22   Pearson chi2:                 1.37e+03
No. Iterations:                    12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.5576      0.102     -5.454      0.0

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:2: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  


Para o modelo sem intercepto, considerando $\alpha=0.05$, a variável private não é significativa.

#### Modelo sem intercepto e variável private

In [ ]:
ajuste44 = smf.glm(formula='y ~ -1 + MEN + URBAN + AGE + SENIORITY', family=sm.families.Binomial(sm.families.links.cloglog),
                  data=base_treino).fit() 
print(ajuste44.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                 1400
Model:                            GLM   Df Residuals:                     1396
Model Family:                Binomial   Df Model:                            3
Link Function:                cloglog   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -783.49
Date:                Thu, 09 Dec 2021   Deviance:                       1567.0
Time:                        23:23:22   Pearson chi2:                 1.37e+03
No. Iterations:                    11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
MEN           -0.5461      0.100     -5.438      0.0

C:\Users\inacio\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Calling Family(..) with a link class as argument is deprecated.
Use an instance of a link class instead.
  """Entry point for launching an IPython kernel.


Para o modelo sem intercepto e sem a variável private, considerando $\alpha=0.05$, todas variáveis são significativas.

#### Comparando o AIC dos modelos ajustados

In [ ]:
print('Modelo completo:', ajuste4.aic,'\n',
      'Modelo sem a variável private:',ajuste42.aic,'\n',
      'Modelo sem intercepto:',ajuste43.aic,'\n',
      'Modelo sem intecepto e sem a variável private:',ajuste44.aic)

Modelo completo: 1578.4059340591489 
 Modelo sem a variável private: 1576.8596284883038 
 Modelo sem intercepto: 1576.6694697599632 
 Modelo sem intecepto e sem a variável private: 1574.9801317485394


Para o modelo com ligação cloglog, segundo o critério AIC, o melhor modelo ajustado é o sem intercepto e sem a variável private.

#### A análise de pontos será realizada para o modelo com menor AIC, que foi o modelo com ligação probito sem a variável private.

In [ ]:
infl = ajuste22.get_influence(observed=False)
summ_df = infl.summary_frame()
summ_df.sort_values("cooks_d", ascending=False)[:10]

dfb_Intercept   dfb_MEN  dfb_URBAN   dfb_AGE  dfb_SENIORITY   cooks_d  \
427       -0.029779  0.008134   0.006104  0.173306      -0.338444  0.024954   
382       -0.107704  0.067315  -0.114824  0.241413      -0.243642  0.019459   
522       -0.035899  0.149598  -0.115082  0.085127      -0.219092  0.014916   
1279      -0.114358  0.048655  -0.096757  0.208820      -0.160064  0.012481   
1208      -0.046945  0.064674  -0.094885  0.150104      -0.188231  0.010768   
65        -0.042561  0.133580  -0.104053  0.075905      -0.175670  0.010761   
512       -0.146492  0.023943  -0.079155  0.199622      -0.068797  0.010216   
508       -0.040912  0.064327  -0.092798  0.140929      -0.182531  0.010075   
847       -0.019244  0.131697  -0.097862  0.048985      -0.166372  0.009931   
1077      -0.045633  0.060748  -0.090275  0.139657      -0.169277  0.009206   

      standard_resid  hat_diag  dffits_internal  
427        -3.459986  0.010315        -0.353226  
382         7.268124  0.001838         0.311923  
522        -3.623233  0.005649        -0.273092  
1279        5.308646  0.002209         0.249807  
1208        5.148181  0.002027         0.232037  
65         -3.111571  0.005527        -0.231959  
512         3.883817  0.003375         0.226008  
508         4.964135  0.002040         0.224445  
847        -2.854340  0.006058        -0.222835  
1077        4.748507  0.002037         0.214545

Aparentemente não há pontos de alavanca e influentes pois não há valores de cook e hat muito altos.

#### d) Qual é o melhor modelo para esses dados na base de treino? Use estatísticas como AIC, curva ROC na base de treino

Como o critério AIC foi utilizado no item c), vamos pegar os modelos sem a variável private para as ligações logito,probito e cauchito e o modelo sem intercepto e sem a variável private para o modelo com ligação cloglog.

Importação da biblioteca para curva ROC

In [ ]:
from sklearn.metrics import roc_curve, auc

Cálculo curva ROC

In [ ]:
yobs = base_treino['y'] 
phat1 = ajuste12.predict(base_treino) 
phat2 = ajuste22.predict(base_treino) 
phat3 = ajuste32.predict(base_treino)
phat4 = ajuste44.predict(base_treino)

fpr1, tpr1, thresholds = roc_curve(yobs, phat1) 
auc1 = auc(fpr1, tpr1) 
 
fpr2, tpr2, thresholds = roc_curve(yobs, phat2) 
auc2 = auc(fpr2, tpr2) 

fpr3, tpr3, thresholds = roc_curve(yobs, phat3) 
auc3 = auc(fpr3, tpr3) 

fpr4, tpr4, thresholds = roc_curve(yobs, phat4) 
auc4 = auc(fpr4, tpr4) 
 
print("Área sob a curva ROC modelo 1 : %f" % auc1) 
print("Área sob a curva ROC modelo 2 : %f" % auc2) 
print("Área sob a curva ROC modelo 3 : %f" % auc2) 
print("Área sob a curva ROC modelo 4 : %f" % auc2) 

Área sob a curva ROC modelo 1 : 0.768910
Área sob a curva ROC modelo 2 : 0.768937
Área sob a curva ROC modelo 3 : 0.768937
Área sob a curva ROC modelo 4 : 0.768937


Os valores da curva ROC para todos modelos foram praticamente iguais, então o melhor modelo será escolhido através do critério AIC

In [ ]:
print("AIC modelo 1 : %f" % ajuste12.aic) 
print("AIC modelo 2 : %f" % ajuste22.aic) 
print("AIC modelo 3 : %f" % ajuste32.aic) 
print("AIC modelo 4 : %f" % ajuste44.aic) 

AIC modelo 1 : 1551.959271
AIC modelo 2 : 1551.827868
AIC modelo 3 : 1566.426660
AIC modelo 4 : 1574.980132


O modelo com menor AIC foi o com ligação probito sem a variável private, logo é o melhor modelo para esses dados na base de treino.

#### e) Considerando os dados de teste, calcule a acurácia do melhor modelo escolhido e d) e compare este resultado com a acurácia do modelo de regressão binaria com ligação logística se for um diferente modelo. Comente. 


Importação da biblioteca para cálculo da acurácia

In [ ]:
from sklearn.metrics import accuracy_score 

Predição de valores na base de teste a cálculo da acurácia

In [ ]:
ypred1 = np.round(ajuste12.predict(base_teste.drop(columns={'y'},axis=1,inplace=False)),0).astype('int64') 
ypred2 = np.round(ajuste22.predict(base_teste.drop(columns={'y'},axis=1,inplace=False)),0).astype('int64')

In [ ]:
ac1 = accuracy_score(ypred1,base_teste['y']) 
ac2 = accuracy_score(ypred2,base_teste['y']) 
print('Acurácia do modelo com ligação logito sem a variável private:',ac1, 
      '\nAcurácia do modelo com ligação probito sem a variável private:',ac2) 

Acurácia do modelo com ligação logito sem a variável private: 0.72 
Acurácia do modelo com ligação probito sem a variável private: 0.7133333333333334


Comparando o melhor modelo escolhido que foi o modelo com ligação probito com o modelo com ligação logito, temos uma acurácia próxima, porém o modelo com ligação logito é melhor para a base de teste.